### DISCUSSION TITLE
Extract the decay products of the Generator level particles

### DISCUSSION THREAD

**Discussion by gsaha009** [columnflow/columnflow#430] [General]
Hi,

I am using `distinctChildrenIdxG` to get the children of a genparticle. it returns some Global indices and then I applied `events.GenPart._apply_global_index(IndicesG)`. After that I could manage to extract the objects from the `GenPart` array. But, this `_apply_global_index` method isn't working in the `production` level. Is there any other way to do this?
As it is a coffea private function, perhaps, it should not be used. Please let me know if anyone has any idea.
Thanks :)

**Comment by pkausw** [comment]
Hi @gsaha009 ,

I think the issue that you see here is that we don't save the `*IdxG` information in the `ReduceEvents` step (those lead to some problems in the past, so we decided to drop them - though after the temporary fix for the `union` arrays, we might want to revisit that topic at some point).

What you should be able to do regardless is to do the matching in e.g. the selection step and save the gen decay products as a new field in the `SelectionResults.objects` section. Since the indices you obtain with e.g. `distinctChildren` are w.r.t. `GenPart` array, this should in principle work out of the box. You could of course also try to save the `*IdxG` information in ReduceEvents by modifying your config - however, as mentioned before, this didn't work in the past and it's not recommended at the moment.

Let us know if this helps!

**Reply by pkausw** [reply]
Ah I forgot, here is also an example of a gen matching that we have implemented for top quarks in case that helps: https://github.com/columnflow/columnflow/blob/master/columnflow/production/cms/gen_top_decay.py

**Comment by gsaha009** [comment]
Hi @pkausw 
Thank you for the answer. I have managed to bypass it for the time being. I will definitely have a thorough look at this. I am closing this discussion.
